jsonschema/tests/test_format.py

```diff
@@ -86,3 +86,100 @@ def test_repr(self):
             repr(checker),
             "<FormatChecker checkers=['bar', 'baz', 'foo']>",
         )
+
+
+class FormatTestMixin:
+    valid_instances = []
+    invalid_instances = []
+    neutral_instances = [
+        None,
+        2,
+        [],
+        {},
+        set(),
+    ]
+    raises = Exception
+    function = None
+
+    @classmethod
+    def call_function(cls, instance):
+        return cls.function(instance)
+
+    def test_valid_instances(self):
+        """
+        Test against a list of valid instances.
+        """
+        for instance in self.valid_instances:
+            self.assertTrue(self.call_function(instance))
+
+    def test_invalid_instances(self):
+        """
+        Test against a list of invalid instances.
+        """
+        for instance in self.invalid_instances:
+            with self.assertRaises(self.raises):
+                self.call_function(instance)
+
+    def test_neutral_instances(self):
+        """
+        Test against a list of types that should be skipped.
+        """
+
+        for instance in self.neutral_instances:
+            self.assertTrue(self.call_function(instance))
+
+
+try:
+    import isoduration
+
+    from jsonschema._format import is_duration
+except ImportError:
+    pass
+else:
+    class TestDurationFormat(TestCase, FormatTestMixin):
+        valid_instances = [
```

Comment:
All these should be moved upstream.

Reply:
I will remove them, I was not sure why the `duration` format was not covered by the tests. Turns out this is related to `pypy3` as this runs on default with Python 3.6 and the library requires Python >=3.7. I changed the gihub action to use Python 3.9 instead. We can also use `pypy-3.7` if pypy is a requirement,

Reply:
Running on 3.9 as you did is fine!

Reply:
Test cases moved to https://github.com/json-schema-org/JSON-Schema-Test-Suite/pull/494